packages/@uppy/companion/src/server/provider/drive/index.js

```diff
@@ -0,0 +1,159 @@
+      this.client
+        .query()
+        .get('teamdrives')
+        .where({ fields: TEAM_DRIVE_FIELDS })
+        .auth(options.token)
+        .request((err, resp) => {
+          if (err) {
+            logger.error(err, 'provider.drive.teamDrive.error')
+          }
+          teamDrivesDone = true
+          teamDrives = resp
+          if (listDone) {
+            finishReq()
+          }
+        })
+    }
+
+    let where = {
+      fields: DRIVE_FILES_FIELDS,
+      q: `'${directory}' in parents and trashed=false`
+    }
+    if (teamDrive) {
+      // Team Drives require several extra parameters in order to work.
+      where.supportsTeamDrives = true
+      where.includeTeamDriveItems = true
+      where.teamDriveId = directory
+      where.corpora = 'teamDrive'
+    }
+
+    this.client
+      .query()
+      .get('files')
+      .where(where)
+      .auth(options.token)
+      .request((err, resp) => {
+        listDone = true
+        listResponse = resp
+        reqErr = err
+        if (teamDrivesDone || directory !== 'root') {
+          finishReq()
+        }
+      })
+  }
+
+  stats ({ id, token }, done) {
+    return this.client
+      .query()
+      .get(`files/${id}`)
+      .where({ fields: DRIVE_FILE_FIELDS, supportsTeamDrives: true })
+      .auth(token)
+      .request(done)
+  }
+
+  download ({ id, token }, onData) {
+    return this.client
+      .query()
+      .get(`files/${id}`)
+      .where({ alt: 'media', supportsTeamDrives: true })
+      .auth(token)
+      .request()
+      .on('data', onData)
+      .on('end', () => onData(null))
+      .on('error', (err) => {
+        logger.error(err, 'provider.drive.download.error')
+      })
+  }
+
+  thumbnail ({id, token}, done) {
+    return this.stats({id, token}, (err, resp, body) => {
+      if (err) {
+        logger.error(err, 'provider.drive.thumbnail.error')
+        return done(null)
+      }
+      done(body.thumbnailLink ? request(body.thumbnailLink) : null)
+    })
+  }
+
+  size ({id, token}, done) {
+    return this.stats({ id, token }, (err, resp, body) => {
+      if (err) {
+        logger.error(err, 'provider.drive.size.error')
+        return done(null)
+      }
+      done(parseInt(body.size))
+    })
+  }
+
+  adaptData (res, teamDrivesResp, uppy) {
+    const data = { username: adapter.getUsername(res), items: [] }
+    const items = adapter.getItemSubList(res)
+    const teamDrives = teamDrivesResp ? teamDrivesResp.teamDrives || [] : []
+    items.concat(teamDrives).forEach((item) => {
+      data.items.push({
+        isFolder: adapter.isFolder(item),
+        icon: adapter.getItemIcon(item),
+        name: adapter.getItemName(item),
+        mimeType: adapter.getMimeType(item),
+        id: adapter.getItemId(item),
+        thumbnail: uppy.buildURL(adapter.getItemThumbnailUrl(item), true),
+        requestPath: adapter.getItemRequestPath(item),
+        modifiedDate: adapter.getItemModifiedDate(item),
+        raw: item
```

Comment:
It looks like `raw` is only used for team drive detection. Could we also do that on the server side instead and add an `isTeamDrive` property to GDrive items? So we don't send unnecessary data to the client.

Reply:
nice! I guess I was restricting myself from sending varying data per Provider. But I think what I can do is something like this:

```js
custom: {
    isTeamDrive: true
}
```
so all custom properties per provider can live there

Reply:
:+1: That looks good to me too

Reply:
updated this now